DOC-1910: Add basic overview of DOMParser and DomPurify use to Security Guide and Migration Guide (#2628)
tinymce/tinymce-docs: modules/ROOT/partials/security/sanitizing-html-input-and-protecting-against-xss-attacks-dom-parser-and-dom-purify.adoc

[[sanitizing-html-input-to-protect-against-xss-attacks]]
=== Sanitizing HTML input to protect against XSS attacks

> Previously, before HTML content was passed to {productname} 5.x, it was parsed using an https://tiny.cloud/docs/api/tinymce.html/tinymce.html.saxparser/[in-house developed API], `SaxParser`.

Reviewer: If we ever get around to finishing that migration of the 5x docs to antora, the link will break.

A link to the branch is also not guaranteed - you'd need to link to a _commit_ tree for that - but I think it's more likely to be valid long-term.

The docs were rendered from the code so all the same information is available.

```suggestion
Previously, before HTML content was passed to {productname} 5.x, it was parsed using an https://github.com/tinymce/tinymce/blob/release/5.10/modules/tinymce/src/core/main/ts/api/html/SaxParser.ts[in-house developed API], `SaxParser`.
```

The `SaxParser` API was developed in the then-absence of alternatives.

When this API’s validate setting was enabled — `validate: true` — `SaxParser` removed elements and attributes that did not fit the declared schema.

And, over its lifetime, `SaxParser` was extended. For example, as of {productname} 5.9, the `SaxParser` API marked attributes with certain names or IDs as unsafe, because some names or IDs can cause the host browser to overwrite existing properties or functions.

For {productname} 6.0, however, this basic parser was removed and replaced with two significantly more thorough alternatives:

. the https://developer.mozilla.org/en-US/docs/Web/API/DOMParser[native browser API], `DOMParser()`; and
. the Free and Open Source _https://github.com/cure53/DOMPurify[XSS sanitizer for HTML, MathML and SVG]_, `DOMPurify`.

> NOTE: {productname} uses DomPurify [2.4.4 | 3.0.0]. <Sentence here explaining why we use the [old | new] version.

Reviewer: Version 3 was published 10 days ago, I haven't even checked to see what's new yet 😛

We currently depend on 2.3.8 because we have custom patches and nobody's tried to upgrade them (or contribute them) since then. But I don't want to be specific.

```suggestion
NOTE: {productname} uses DomPurify 2.x, which was current at the time version 6 was developed.
```

Reviewer: Looks like the main change is dropping support for protecting against exploits that only work on IE. I'll make a note to consider upgrading, as this sounds like it will reduce our bundle size...

Before HTML (or XML) content is passed to {productname} 6.x, the `DOMParser` API parses the HTML (or XML) string into a https://developer.mozilla.org/en-US/docs/Web/API/Document_Object_Model[DOM object]. As part of this process, `DOMParser` attempts to correct mal-formed HTML.

For example, the following string:

[source,html]
----
<p>
<a href="https://example.com">Click here.
</p>
----

When this string is passed in, the `DOMParser` API adds the missing closing anchor tag:

[source,html]
----
<p>
<a href="https://example.com">Click here.</a>
</p>
----

Next, the, now correctly formed, DOM object is passed to `DOMPurify`.

`DOMPurify` runs a security-focussed ruleset over the DOM object to sanitize the HTML and protect {productname}, and applications built using {productname}, against XSS attacks.

==== Turning `DOMPurify` off

> Both `DOMParser` and `DOMPurify` are on by default, and `DOMParser` cannot be turned off.

Reviewer 2: `DOMParser` is an essential part of our parsing toolchain, we should not mention the idea of turning it off.

> As of {productname} 6.4, however, it is possible to turn `DOMPurify` off using the `xss_sanitization` option.

Reviewer 2: ```suggestion
As of {productname} 6.4, however, it is possible to turn DOMPurify off using the `xss_sanitization` option.
```

NOTE: `xss_sanitization` is set to `true` by default. That is, {productname} specifically sets the `xss_sanitization` option to `true`, even if this option is not declared in a {productname} configuration.

*Type:* `+Boolean+`

*Default value:* `+true+`

*Possible values:* `+true+`, `+false+`

===== Example: using `+xss_sanitization+` to turn DOMPurify off and no longer sanitize HTML against XSS attacks

[source,js]
----
tinymce.init({
  selector: 'textarea',  // change this value according to your HTML
  xss_sanitization: false
});
----

> WARNING: Turning `DomPurify` off leaves {productname}, and any application using {productname}, extremely vulnerable to XSS attacks. Only turn `DomPurify` off if alternative and equivalently capable HTML and XML sanitisation and XSS protections are in place.

Author: Thanks to the vagaries of GitHub’s UI, it ended up being easier to make this change remotely. So the easier path was taken.
